src/input/composite_device/mod.rs

```diff
@@ -1152,105 +1152,57 @@ impl CompositeDevice {
             }
         }
 
-        for mut event in events {
+        for event in events {
             let cap = event.as_capability();
 
             // Track what is currently active so we can ignore extra events.
-            match cap {
+            match cap.clone() {
                 Capability::None
                 | Capability::NotImplemented
                 | Capability::Sync
                 | Capability::DBus(_) => {}
-                Capability::Gamepad(_) | Capability::Mouse(_) | Capability::Keyboard(_) => {
+                Capability::Keyboard(_) => {
                     if !self.set_active_events(&cap, is_pressed) {
                         continue;
                     }
-                }
-            }
-            // Process the event depending on the intercept mode
-            // Check if we have met the criteria for InterceptMode:Always
-            if intercept && self.intercept_activation_caps.contains(&cap) {
-                log::debug!("Found matching intercept event: {:?}", cap);
-                // TODO: This assumes the list is unique. Switch to HashSet
-                if is_pressed && self.should_hold_intercept_input(&cap) {
-                    // Stop here if this is a repeat event.
-                    if self.intercept_active_inputs.contains(&cap) {
-                        log::debug!("The event is already in the list. Skipping.");
-                        continue;
-                    };
-                    self.intercept_active_inputs.push(cap.clone());
-
-                    if self.intercept_active_inputs.len() != self.intercept_activation_caps.len() {
-                        log::debug!("More events needed to activate intercept mode.");
+                    if !self
+                        .check_intercept_keys(event.clone(), is_pressed, intercept)
```

Comment:
We should borrow `event` here instead of cloning it to avoid cloning all button events.